File: py-basemapper/src/basemapper/layer_utils.py
"""Layer discovery utility."""

from __future__ import annotations

import json
from typing import List


def list_layers(style_input: str) -> List[str]:
    """Return all layer IDs from a MapLibre GL style.

    Also prints a two-column table of ``id`` and ``type`` to stdout for
    interactive exploration.

    Args:
        style_input: A MapLibre GL style URL or inline JSON string.

    Returns:
        Sorted list of layer ``id`` values.

    Raises:
        BasemapError: If the style cannot be fetched or parsed.
    """
    from .exceptions import BasemapError

    if style_input.strip().startswith("{"):
        try:
            style = json.loads(style_input)
        except json.JSONDecodeError as exc:
            raise BasemapError(f"style parse error: {exc}") from exc
    else:
        import requests

        try:
            resp = requests.get(style_input, timeout=10)
            resp.raise_for_status()
            style = resp.json()
        except Exception as exc:
            raise BasemapError(f"style fetch failed: {exc}") from exc

    layers = style.get("layers", [])
    ids = [layer.get("id", "") for layer in layers if "id" in layer]

    # Print interactive table.
    print(f"{'id':<40} {'type':<20}")
    print("-" * 61)
    for layer in layers:
        lid = layer.get("id", "")
        ltype = layer.get("type", "")
        print(f"{lid:<40} {ltype:<20}")

    return sorted(ids)

File: r-basemapper/R/layer_utils.R
#' List all layer IDs from a MapLibre GL style.
#'
#' Fetches (if URL) or parses (if inline JSON) a MapLibre GL style and returns
#' all layer `id` values. Also prints a two-column data.frame of id and type to
#' the console for interactive exploration.
#'
#' @param style_input Character: a MapLibre GL style URL or inline JSON string.
#' @return A character vector of all layer `id` values.
#' @export
#' @examples
#' \dontrun{
#' ids <- list_layers("https://demotiles.maplibre.org/style.json")
#' }
list_layers <- function(style_input) {
  style_json <- if (startsWith(trimws(style_input), "{")) {
    style_input
  } else {
    resp <- httr2::request(style_input) |>
      httr2::req_timeout(10) |>
      httr2::req_perform()
    httr2::resp_body_string(resp)
  }

  style <- jsonlite::fromJSON(style_json, simplifyVector = FALSE)
  layers <- style[["layers"]]

  if (is.null(layers) || length(layers) == 0) {
    message("No layers found in style.")
    return(character(0))
  }

  ids   <- vapply(layers, function(l) l[["id"]]   %||% "", character(1))
  types <- vapply(layers, function(l) l[["type"]] %||% "", character(1))

  df <- data.frame(id = ids, type = types, stringsAsFactors = FALSE)
  print(df)

  invisible(sort(ids))
}
